### PR TITLE
New Data Dictionary framework and module

### PR DESCRIPTION
**Pull request by driusan**
This updates the LORIS Data Framework to include a concept of a data dictionary that a module may manage. It adds a new "dictionary" module to manage the module dictionary but leaves the `datadict` module in place because it's currently the only place to manage the `parameter_type` table used by the DQT.

The extension of the data framework to include dictionaries also includes many enhancements from the current "data dictionary" in LORIS which only includes a field name and description. Types, Scopes, and Cardinalities are added which more thoroughly describe the data and will eventually be able to be used to build richer data-type sensitive interfaces in places like the DQT. DictionaryItems implement the AccessibleResource interface to enable modules to give column-level granularity over data.

Modules provide categories of DictionaryItems, which allows them to group items in the frontend. This is primarily to make it possible to filter the instruments from the "instruments" module (in the current `data_dict` module, the `Source From` options are literally populated by a call to `\Utility::getAllInstruments()`) in a manageable way, but can be used by any module that wants to group the fields.

Instruments written in PHP must now implement a new `getDataDictionary()` abstract function to describe the fields provided by the instrument, but a `LorisFormDictionaryImpl` is provided which populates it in the same way that the `lorisform_parser` uses it to populate the `parameter_type` table. The `candidate_parameters` and `imaging_browser` also implement dictionaries in this PR.

NB: I intend to try to break this PR into smaller, easier to review chunks, but it may be helpful to reviewers to have a PR with the entire dictionary overhaul for context.

### CODE REVIEW

**Comment by driusan** [comment]
@HenriRabalais can you have a look at this to see if it's possible to describe the biobank's data? The place to start would probably be the `$module->getDataDictionary()` function and some of the examples of modules that implement it.
@zaliqarosli can you look at it to see if it's missing anything (in particular types) wrt. the work you've done on instruments that would be needed?

**Comment by driusan** [comment]
Blocked by #6938, #6939

File: modules/candidate_parameters/php/module.class.inc
Comment (xlecours) [inline modules/candidate_parameters/php/module.class.inc:226]: There is no constraint on PSCID in our database schema. This should be SINGLE because there should be one value per candidate; right?

File: modules/dictionary/php/fields.class.inc
Comment (xlecours) [inline modules/dictionary/php/fields.class.inc:13]: no need for NDB_Page. `implements RequestHandlerInetrface` would be enough.

File: modules/dictionary/php/fields.class.inc
Comment (xlecours) [inline modules/dictionary/php/fields.class.inc:63]: I would go with 202 since 201 do not fit the situation when an override is deleted.

File: modules/candidate_parameters/php/module.class.inc
Comment (driusan) [reply modules/candidate_parameters/php/module.class.inc:226]: @ridz1208 it looks like he's right.. in the SQL schema the PSCID is defined as `KEY `PSCID` (`PSCID`),` not `UNIQUE KEY`. Do you have any idea why that is? It doesn't seem right to me.

File: modules/dictionary/php/fields.class.inc
Comment (xlecours) [inline modules/dictionary/php/fields.class.inc:81]: Considering the amount of usage this function will get, take the following lightly...
I think keeping the original_value in the database would reduce the amount of operations required to check if the override is set to the original value.

**Review by xlecours — DISMISSED** [review]

**Comment by zaliqarosli** [comment]
@driusan can you point me to where i should be looking? this is a big PR and i don't know what it is exactly you want me to look at to see if something is missing. and what kind of 'types' are you referring to? thanks

**Comment by driusan** [comment]
The `src/Data` directory parts of this PR which is separated out into PR #6939

File: modules/candidate_parameters/php/module.class.inc
Comment (ridz1208) [reply modules/candidate_parameters/php/module.class.inc:226]: maybe to accomodate scanners ? or non-human entities ? I think UNIQUE KEY prevents multiple values from being NULL also. is the PSCID NOT NULL in the DB ?

**Comment by driusan** [comment]
Multiselects should have multiple cardinality. (See comment on #6939)

**Comment by driusan** [comment]
Now blocked by #7176, #7177,  and #7179

File: modules/candidate_parameters/php/module.class.inc
Comment (driusan) [reply modules/candidate_parameters/php/module.class.inc:226]: Addressed in #7176

File: modules/dictionary/php/fields.class.inc
Comment (driusan) [reply modules/dictionary/php/fields.class.inc:63]: But this is in the case where it's handling a PUT request?

Comment (driusan) [reply modules/dictionary/php/fields.class.inc:63]: Oh, I see. `_replaceDescription` can delete. (It's been a while since I looked at this code..)

File: modules/dictionary/php/moduledict.class.inc
Comment (driusan) [reply modules/dictionary/php/moduledict.class.inc:40]: I added a check to ensure it's a GET.

File: modules/dictionary/php/fields.class.inc
Comment (driusan) [reply modules/dictionary/php/fields.class.inc:81]: I don't think it's worth putting in the database, it's not causing a bottleneck and that would tightly couple it to the DB.

File: modules/dictionary/php/fields.class.inc
Comment (driusan) [reply modules/dictionary/php/fields.class.inc:13]: I don't see what the benefit of that would be.

**Comment by driusan** [comment]
I've rebased this now that the blocking dictionaries are merged and I think I've responded to all outstanding reviews.. I think this is ready for someone to review the module itself. (Still blocked by #7180, but I don't foresee any significant changes there that would hold back reviewing the module..)

**Comment by driusan** [comment]
@maltheism @PapillonMcGill can you guys review this as discussed in the LORIS meeting?

**Comment by maltheism** [comment]
Hi @driusan, I'm getting 404 when visiting `/dictionary` after refresh of rasinbread_refresh.php. I now realize there isn't sql patches in this PR. What are the setup and testing instructions for manually testing?

**Comment by driusan** [comment]
@maltheism I think it's just not in the modules table because I forgot to add a patch..  the script`manage_modules.php --add` from the tools directory should also work to test it before I get around to adding an SQL patch.

File: modules/dictionary/jsx/dataDictIndex.js
Comment (maltheism) [inline modules/dictionary/jsx/dataDictIndex.js:88]: Hi @driusan, spelling mistake for asynchronously.

File: modules/dictionary/jsx/dataDictIndex.js
Comment (maltheism) [inline modules/dictionary/jsx/dataDictIndex.js:151]: @driusan spelling mistake for the word Retrieve.

File: modules/dictionary/jsx/dataDictIndex.js
Comment (maltheism) [inline modules/dictionary/jsx/dataDictIndex.js:175]: @driusan spelling mistake for the word formatted.

**Comment by maltheism** [comment]
Hi @driusan, everything works for me from manually testing.

**Comment by driusan** [comment]
@maltheism I think I fixed all the things you pointed out.

**Comment by driusan** [comment]
@maltheism just realized I forgot about the SQL patch.. added now.

**Review by maltheism — APPROVED** [review]
Hi @driusan, everything in the PR looks good to me. One thing to note is that the `./raisinbread_refresh.php` script in tools, didn't add your SQL changes unless I emptied my database and ran the installdb.php first. I think that's the norm until a release patch.

**Comment by driusan** [comment]
@xlecours can you update or dismiss your review?

**Comment by maltheism** [comment]
@driusan needs rebase

**Comment by driusan** [comment]
@maltheism rebase done

@xlecours my previous comment seems to be hidden since rebasing.. could you update or dismiss your review? I think I've responded to everything in it.

File: modules/dictionary/php/fields.class.inc
Comment (xlecours) [reply modules/dictionary/php/fields.class.inc:13]: Better cohesion and reduced coupling. If you only need the handle function, and maybe _hasAccess, implementing  RequestHandlerInterface is enough. I don't see what $identifer and $commentid means for a datadictionnary field. $fieldoptions, $dateoption adds confusion. What is the use of AddFile, AddSelect, ...? Will it use its LorisForm?

**Comment by xlecours** [comment]
> @maltheism rebase done
> 
> @xlecours my previous comment seems to be hidden since rebasing.. could you update or dismiss your review? I think I've responded to everything in it.

I have added a comment there: https://github.com/aces/Loris/pull/6936#discussion_r616869032
YAY? NAY? Just let me know, I'll approve or review the changes accordingly.

File: modules/dictionary/php/fields.class.inc
Comment (driusan) [reply modules/dictionary/php/fields.class.inc:13]: To be honest I'd like to remove `$identifier` and `$commentid` from the page class in general, but when I tried doing that in #7414 the scope of the change exploded so I held back to do it later.

I don't think having methods that are uncalled adds any overhead and the properties in the page class add minimal. I just had a look through the code and hasAccess does seem to be the main benefit. Other than that, I think there's benefit in having a root to the class hierarchy, but I can change this to manually call hasAccess if you want.

Comment (driusan) [reply modules/dictionary/php/fields.class.inc:13]: I made #7431 to work through what would be ideal in general outside of this PR.

**Pull request merged**
